scripts/exporters/gltf-exporter.js

```diff
@@ -0,0 +1,441 @@
+                }
+            });
+        };
+
+        resources.entities.forEach((entity) => {
+            if (entity.camera) {
+                resources.cameras.push(entity.camera);
+            }
+
+            if (entity.render) {
+                collectMaterials(entity.render.meshInstances);
+            }
+
+            if (entity.model) {
+                collectMaterials(entity.model.meshInstances);
+            }
+        });
+
+        return resources;
+    }
+
+    writeBuffers(resources, json) {
+        if (resources.buffers.length > 0) {
+            json.buffers = [];
+
+            let byteLength = 0;
+
+            resources.buffers.forEach((buffer) => {
+                const arrayBuffer = buffer.lock();
+                byteLength += arrayBuffer.byteLength;
+            });
+
+            const buffer = {
+                byteLength: byteLength
+            };
+
+            json.buffers.push(buffer);
+        }
+    }
+
+    writeBufferViews(resources, json) {
+        if (resources.buffers.length > 0) {
+            json.bufferViews = [];
+
+            let offset = 0;
+
+            resources.buffers.forEach((buffer) => {
+                const arrayBuffer = buffer.lock();
+
+                const bufferView = {
+                    buffer: 0,
+                    byteLength: arrayBuffer.byteLength,
+                    byteOffset: offset
+                };
+
+                if (buffer instanceof pc.VertexBuffer) {
+                    bufferView.target = ARRAY_BUFFER;
+                    bufferView.byteStride = buffer.getFormat().size;
+                } else {
+                    bufferView.target = ELEMENT_ARRAY_BUFFER;
+                }
+
+                offset += arrayBuffer.byteLength;
+
+                json.bufferViews.push(bufferView);
+            });
+        }
+    }
+
+    writeCameras(resources, json) {
+        if (resources.cameras.length > 0) {
+            json.cameras = [];
+
+            resources.cameras.forEach((cam) => {
+                const projection = cam.projection;
+                const nearClip = cam.nearClip;
+                const farClip = cam.farClip;
+                const fov = cam.fov;
+
+                if (projection === pc.PROJECTION_ORTHOGRAPHIC) {
+                    json.cameras.push({
+                        type: "orthographic",
+                        orthographic: {
+                            xmag: 1,
+                            ymag: 1,
+                            znear: nearClip,
+                            zfar: farClip
+                        }
+                    });
+                } else {
+                    json.cameras.push({
+                        type: "perspective",
+                        perspective: {
+                            yfov: fov * Math.PI / 180,
+                            znear: nearClip,
+                            zfar: farClip
+                        }
+                    });
+                }
+            });
```

Comment:
A `resources.cameras.forEach` combined with `push` in every case is basically a `resources.cameras.map`, which may give that code more intention (probably subjective, I just see it as less manual array plumbing)

Reply:
Good point, yeah, I don't disagree.

Reply:
OK, I've updated most `forEach`+`push` combinations to use `map` instead. Nice suggestion.